File: java/client/src/main/java/org/signal/libsignal/net/ChatService.java
//
// Copyright 2024 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

package org.signal.libsignal.net;

import java.lang.ref.WeakReference;
import java.net.MalformedURLException;
import java.util.Map;
import java.util.function.Function;
import org.signal.libsignal.internal.CalledFromNative;
import org.signal.libsignal.internal.CompletableFuture;
import org.signal.libsignal.internal.FilterExceptions;
import org.signal.libsignal.internal.Native;
import org.signal.libsignal.internal.NativeHandleGuard;
import org.signal.libsignal.net.internal.BridgeChatListener;

/** Represents an API of communication with the Chat Service. */
public abstract class ChatService extends NativeHandleGuard.SimpleOwner {
  private final TokioAsyncContext tokioAsyncContext;
  protected final ChatListener chatListener;

  ChatService(
      final TokioAsyncContext tokioAsyncContext,
      final Network.ConnectionManager connectionManager,
      Function<Long, Long> chatServiceCreateWrapper,
      ChatListener chatListener) {
    super(connectionManager.guardedMap(chatServiceCreateWrapper::apply));
    this.tokioAsyncContext = tokioAsyncContext;
    this.chatListener = chatListener;

    if (chatListener != null) {
      var listenerBridge = new ListenerBridge(this);
      tokioAsyncContext.guardedRun(
          asyncContextHandle ->
              guardedRun(
                  chatServiceHandle ->
                      setListenerWrapper(asyncContextHandle, chatServiceHandle, listenerBridge)));
    }
  }

  private static final class ListenerBridge implements BridgeChatListener {
    // Stored as a weak reference because otherwise we'll have a reference cycle:
    // - After setting a listener, Rust has a GC GlobalRef to this ListenerBridge
    // - This field is a normal Java reference to the ChatService
    // - ChatService owns the Rust ChatService object
    private final WeakReference<ChatService> chat;

    ListenerBridge(ChatService chat) {
      this.chat = new WeakReference<>(chat);
    }

    public void onIncomingMessage(
        byte[] envelope, long serverDeliveryTimestamp, long sendAckHandle) {
      ChatService chat = this.chat.get();
      if (chat == null) return;
      if (chat.chatListener == null) return;

      chat.chatListener.onIncomingMessage(
          chat,
          envelope,
          serverDeliveryTimestamp,
          new ChatListener.ServerMessageAck(chat.tokioAsyncContext, sendAckHandle));
    }

    public void onQueueEmpty() {
      ChatService chat = this.chat.get();
      if (chat == null) return;
      if (chat.chatListener == null) return;

      chat.chatListener.onQueueEmpty(chat);
    }

    public void onConnectionInterrupted(Throwable disconnectReason) {
      ChatService chat = this.chat.get();
      if (chat == null) return;
      if (chat.chatListener == null) return;

      ChatServiceException disconnectReasonChatServiceException =
          (disconnectReason instanceof ChatServiceException)
              ? (ChatServiceException) disconnectReason
              : new ChatServiceException("OtherDisconnectReason", disconnectReason);
      chat.chatListener.onConnectionInterrupted(chat, disconnectReasonChatServiceException);
    }
  }

  /**
   * Initiates termination of the underlying connection to the Chat Service. After the service is
   * disconnected, it will not attempt to automatically reconnect until you call {@link #connect()}.
   *
   * <p>Note: the same instance of {@code ChatService} can be reused after {@code disconnect()} was
   * called.
   *
   * @return a future that completes when the underlying connection is terminated.
   */
  @SuppressWarnings("unchecked")
  public CompletableFuture<Void> disconnect() {
    return tokioAsyncContext.guardedMap(
        asyncContextHandle ->
            guardedMap(
                chatServiceHandle -> disconnectWrapper(asyncContextHandle, chatServiceHandle)));
  }

  /**
   * Initiates establishing of the underlying connection to the Chat Service. Once the service is
   * connected, all the requests will be using the established connection.
   *
   * <p>The resulting future may fail with {@link AppExpiredException}, {@link
   * DeviceDeregisteredException}, or {@link RetryLaterException} (inside an {@link
   * java.util.concurrent.ExecutionException ExecutionException}), along with other {@link
   * ChatServiceException}s.
   *
   * @return a future with the result of the connection attempt (either a {@link DebugInfo} or an
   *     error).
   */
  public CompletableFuture<DebugInfo> connect() {
    return tokioAsyncContext.guardedMap(
        asyncContextHandle ->
            guardedMap(
                chatServiceHandle ->
                    connectWrapper(asyncContextHandle, chatServiceHandle)
                        .thenApply(o -> (DebugInfo) o)));
  }

  /**
   * Sends request to the Chat Service over this channel.
   *
   * @param req request object
   * @return a {@code CompletableFuture} of a {@link Response}. The future will fail with a {@link
   *     ChatServiceInactiveException} (inside an {@link java.util.concurrent.ExecutionException
   *     ExecutionException}) if you haven't called {@link #connect()} }.
   * @throws MalformedURLException if {@code pathAndQuery} component of the request has an invalid
   *     structure.
   */
  public CompletableFuture<Response> send(final Request req) throws MalformedURLException {
    final InternalRequest internalRequest = buildInternalRequest(req);
    try (final NativeHandleGuard asyncContextHandle = new NativeHandleGuard(tokioAsyncContext);
        final NativeHandleGuard chatServiceHandle = new NativeHandleGuard(this);
        final NativeHandleGuard requestHandle = new NativeHandleGuard(internalRequest)) {
      return sendWrapper(
              asyncContextHandle.nativeHandle(),
              chatServiceHandle.nativeHandle(),
              requestHandle.nativeHandle(),
              req.timeoutMillis)
          .thenApply(o -> (Response) o);
    }
  }

  /**
   * Sends request to the Chat Service over this channel.
   *
   * <p>In addition to the response, an object containing debug information about the request flow
   * is returned.
   *
   * @param req request object
   * @return a {@code CompletableFuture} of a {@link ResponseAndDebugInfo}. The future will fail
   *     with a {@link ChatServiceInactiveException} (inside an {@link
   *     java.util.concurrent.ExecutionException ExecutionException}) if you haven't called {@link
   *     #connect()}.
   * @throws MalformedURLException if {@code pathAndQuery} component of the request has an invalid
   *     structure.
   */
  public CompletableFuture<ResponseAndDebugInfo> sendAndDebug(final Request req)
      throws MalformedURLException {
    final InternalRequest internalRequest = buildInternalRequest(req);
    try (final NativeHandleGuard asyncContextHandle = new NativeHandleGuard(tokioAsyncContext);
        final NativeHandleGuard chatServiceHandle = new NativeHandleGuard(this);
        final NativeHandleGuard requestHandle = new NativeHandleGuard(internalRequest)) {
      return sendAndDebugWrapper(
              asyncContextHandle.nativeHandle(),
              chatServiceHandle.nativeHandle(),
              requestHandle.nativeHandle(),
              req.timeoutMillis)
          .thenApply(o -> (ResponseAndDebugInfo) o);
    }
  }

  // These are meant to be thin wrappers around the correct call to Native.ChatService_* calls
  //   for each of the concrete implementing classes.
  protected abstract CompletableFuture disconnectWrapper(
      long nativeAsyncContextHandle, long nativeChatServiceHandle);

  protected abstract CompletableFuture<Object> sendWrapper(
      long nativeAsyncContextHandle,
      long nativeChatServiceHandle,
      long nativeRequestHandle,
      int timeoutMillis);

  protected abstract CompletableFuture<Object> connectWrapper(
      long nativeAsyncContextHandle, long nativeChatServiceHandle);

  protected abstract CompletableFuture<Object> sendAndDebugWrapper(
      long nativeAsyncContextHandle,
      long nativeChatServiceHandle,
      long nativeRequestHandle,
      int timeoutMillis);

  protected abstract void setListenerWrapper(
      long nativeAsyncContextHandle,
      long nativeChatServiceHandle,
      BridgeChatListener bridgeChatListener);

  static InternalRequest buildInternalRequest(final Request req) throws MalformedURLException {
    final InternalRequest result =
        new InternalRequest(req.method(), req.pathAndQuery(), req.body());
    req.headers().forEach(result::addHeader);
    return result;
  }

  static class InternalRequest extends NativeHandleGuard.SimpleOwner {
    InternalRequest(final String method, final String pathAndQuery, final byte[] body)
        throws MalformedURLException {
      super(
          FilterExceptions.filterExceptions(
              MalformedURLException.class,
              () -> Native.HttpRequest_new(method, pathAndQuery, body)));
    }

    @Override
    protected void release(final long nativeHandle) {
      Native.HttpRequest_Destroy(nativeHandle);
    }

    public void addHeader(final String name, final String value) {
      guardedRun(h -> Native.HttpRequest_add_header(h, name, value));
    }
  }

  public record Request(
      String method,
      String pathAndQuery,
      Map<String, String> headers,
      byte[] body,
      int timeoutMillis) {}

  public record Response(int status, String message, Map<String, String> headers, byte[] body) {}

  public record DebugInfo(IpType ipType, int durationMs, String connectionInfo) {
    @CalledFromNative
    DebugInfo(byte ipTypeCode, int durationMs, String connectionInfo) {
      this(IpType.values()[ipTypeCode], durationMs, connectionInfo);
    }
  }

  public record ResponseAndDebugInfo(Response response, DebugInfo debugInfo) {}
}

File: swift/Sources/LibSignalClient/ChatService.swift
//
// Copyright 2024 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

import Foundation
import SignalFfi

/// Common interface for ongoing communication channels with the Chat Service.
public protocol ChatService: AnyObject {
    /// Initiates establishing of the underlying authenticated connection to the Chat Service. Once
    /// the service is connected, all the requests will be using the established connection. Also,
    /// if the connection is lost for any reason other than the call to ``disconnect()``, an
    /// automatic reconnect attempt will be made.
    ///
    /// Specific subclasses may have more information on the types of errors produced by `connect()`,
    /// as well as additional preconditions that should be satisfied.
    @discardableResult
    func connect() async throws -> DebugInfo

    /// Initiates termination of the underlying connection to the Chat Service. After the service is
    /// disconnected, it will not attempt to automatically reconnect until you call ``connect()``.
    ///
    /// Note: the same `ChatService` can be reused after `disconnect()` was called.
    ///
    /// Returns when the disconnection is complete.
    func disconnect() async throws

    /// Sends a request to the Chat Service.
    ///
    /// - Throws: ``SignalError/chatServiceInactive(_:)`` if you haven't called ``connect()``
    /// - Throws: Other ``SignalError``s for other kinds of failures.
    /// - SeeAlso: ``send(_:)``
    func send(_ request: Request) async throws -> Response

    /// Sends a request to the Chat Service.
    ///
    /// In addition to the response, an object containing debug information about the request flow
    /// is returned.
    ///
    /// - Throws: ``SignalError/chatServiceInactive(_:)`` if you haven't called ``connect()``
    /// - Throws: Other ``SignalError``s for other kinds of failures.
    /// - SeeAlso: ``send(_:)``
    func sendAndDebug(_ request: Request) async throws -> (Response, DebugInfo)
}

extension ChatService {
    public typealias Request = ChatRequest
    public typealias Response = ChatResponse
    public typealias DebugInfo = ChatServiceDebugInfo
}

/// Represents an API of authenticated communication with the Chat Service.
///
/// An instance of this object is obtained via call to ``Net/createAuthenticatedChatService(username:password:receiveStories:)``.
public class AuthenticatedChatService: NativeHandleOwner, ChatService {
    internal let tokioAsyncContext: TokioAsyncContext

    internal init(tokioAsyncContext: TokioAsyncContext, connectionManager: ConnectionManager, username: String, password: String, receiveStories: Bool) {
        var handle: OpaquePointer?
        connectionManager.withNativeHandle { connectionManager in
            failOnError(signal_chat_service_new_auth(&handle, connectionManager, username, password, receiveStories))
        }
        self.tokioAsyncContext = tokioAsyncContext
        super.init(owned: handle!)
    }

    internal required init(owned handle: OpaquePointer) {
        fatalError("should not be called directly for a ChatService")
    }

    override internal class func destroyNativeHandle(_ handle: OpaquePointer) -> SignalFfiErrorRef? {
        return signal_auth_chat_destroy(handle)
    }

    /// Sets (or clears) the listener for server push messages.
    ///
    /// Takes ownership of the listener; be careful this doesn't lead to a reference cycle (unless the owner lives forever anyway).
    public func setListener(_ listener: (any ChatListener)?) {
        self.tokioAsyncContext.withNativeHandle { tokioAsyncContext in
            withNativeHandle { chatService in
                if let listener {
                    var listenerStruct = ChatListenerBridge(chatService: self, chatListener: listener).makeListenerStruct()
                    failOnError(signal_chat_service_set_listener_auth(tokioAsyncContext, chatService, &listenerStruct))
                } else {
                    failOnError(signal_chat_service_set_listener_auth(tokioAsyncContext, chatService, nil))
                }
            }
        }
    }

    /// Initiates establishing of the underlying authenticated connection to the Chat Service. Once
    /// the service is connected, all the requests will be using the established connection. Also,
    /// if the connection is lost for any reason other than the call to ``disconnect()``, an
    /// automatic reconnect attempt will be made.
    ///
    /// Calling this method will result in starting to accept incoming requests from the Chat
    /// Service. You should set a listener first using ``setListener(_:)``.
    ///
    /// - Throws: ``SignalError/appExpired(_:)`` if the current app version is too old (as judged by
    ///   the server).
    /// - Throws: ``SignalError/deviceDeregistered(_:)`` if the current device has been deregistered
    ///   or delinked.
    /// - Throws: ``SignalError/rateLimitedError(_:, _:)`` if the server
    ///   response indicates the request should be tried again after some time.
    /// - Throws: Other ``SignalError``s for other kinds of failures.
    @discardableResult
    public func connect() async throws -> DebugInfo {
        let rawDebugInfo = try await self.tokioAsyncContext.invokeAsyncFunction { promise, tokioAsyncContext in
            withNativeHandle { chatService in
                signal_chat_service_connect_auth(promise, tokioAsyncContext, chatService)
            }
        }
        return DebugInfo(consuming: rawDebugInfo)
    }

    /// Initiates termination of the underlying connection to the Chat Service. After the service is
    /// disconnected, it will not attempt to automatically reconnect until you call ``connect()``.
    ///
    /// Note: the same instance of `AuthenticatedChatService` can be reused after `disconnect()` was
    /// called.
    ///
    /// Returns when the disconnection is complete.
    public func disconnect() async throws {
        _ = try await self.tokioAsyncContext.invokeAsyncFunction { promise, tokioAsyncContext in
            withNativeHandle { chatService in
                signal_chat_service_disconnect_auth(promise, tokioAsyncContext, chatService)
            }
        }
    }

    /// Sends request to the Chat Service over an authenticated channel.
    ///
    /// - Throws: ``SignalError/chatServiceInactive(_:)`` if you haven't called ``connect()``
    /// - Throws: Other ``SignalError``s for other kinds of failures.
    /// - SeeAlso: ``sendAndDebug(_:)``
    public func send(_ request: Request) async throws -> Response {
        let internalRequest = try Request.InternalRequest(request)
        let timeoutMillis = request.timeoutMillis
        let rawResponse: SignalFfiChatResponse = try await self.tokioAsyncContext.invokeAsyncFunction { promise, tokioAsyncContext in
            withNativeHandle { chatService in
                internalRequest.withNativeHandle { request in
                    signal_chat_service_auth_send(promise, tokioAsyncContext, chatService, request, timeoutMillis)
                }
            }
        }
        return try Response(consuming: rawResponse)
    }

    /// Sends request to the Chat Service over an authenticated channel.
    ///
    /// In addition to the response, an object containing debug information about the request flow
    /// is returned.
    ///
    /// - Throws: ``SignalError/chatServiceInactive(_:)`` if you haven't called ``connect()``
    /// - Throws: Other ``SignalError``s for other kinds of failures.
    /// - SeeAlso: ``send(_:)``
    public func sendAndDebug(_ request: Request) async throws -> (Response, DebugInfo) {
        let internalRequest = try Request.InternalRequest(request)
        let timeoutMillis = request.timeoutMillis
        let rawResponse: SignalFfiResponseAndDebugInfo = try await self.tokioAsyncContext.invokeAsyncFunction { promise, tokioAsyncContext in
            withNativeHandle { chatService in
                internalRequest.withNativeHandle { request in
                    signal_chat_service_auth_send_and_debug(promise, tokioAsyncContext, chatService, request, timeoutMillis)
                }
            }
        }
        return (try Response(consuming: rawResponse.response), DebugInfo(consuming: rawResponse.debug_info))
    }
}

/// Represents an API of unauthenticated communication with the Chat Service.
///
/// An instance of this object is obtained via call to ``Net/createUnauthenticatedChatService()``.
public class UnauthenticatedChatService: NativeHandleOwner, ChatService {
    internal let tokioAsyncContext: TokioAsyncContext

    internal init(tokioAsyncContext: TokioAsyncContext, connectionManager: ConnectionManager) {
        var handle: OpaquePointer?
        connectionManager.withNativeHandle { connectionManager in
            failOnError(signal_chat_service_new_unauth(&handle, connectionManager))
        }
        self.tokioAsyncContext = tokioAsyncContext
        super.init(owned: handle!)
    }

    internal required init(owned handle: OpaquePointer) {
        fatalError("should not be called directly for a ChatService")
    }

    override internal class func destroyNativeHandle(_ handle: OpaquePointer) -> SignalFfiErrorRef? {
        return signal_unauth_chat_destroy(handle)
    }

    /// Sets (or clears) the listener for connection events.
    ///
    /// Takes ownership of the listener; be careful this doesn't lead to a reference cycle (unless the owner lives forever anyway).
    public func setListener(_ listener: (any ConnectionEventsListener<UnauthenticatedChatService>)?) {
        self.tokioAsyncContext.withNativeHandle { tokioAsyncContext in
            withNativeHandle { chatService in
                if let listener {
                    var listenerStruct = UnauthConnectionEventsListenerBridge(chatService: self, listener: listener).makeListenerStruct()
                    failOnError(signal_chat_service_set_listener_unauth(tokioAsyncContext, chatService, &listenerStruct))
                } else {
                    failOnError(signal_chat_service_set_listener_unauth(tokioAsyncContext, chatService, nil))
                }
            }
        }
    }

    /// Initiates establishing of the underlying unauthenticated connection to the Chat Service. Once
    /// the service is connected, all the requests will be using the established connection. Also,
    /// if the connection is lost for any reason other than the call to ``disconnect()``, an
    /// automatic reconnect attempt will be made.
    ///
    /// - Throws: ``SignalError/appExpired(_:)`` if the current app version is too old (as judged by
    ///   the server).
    /// - Throws: ``SignalError/rateLimitedError(_:, _:)`` if the server
    ///   response indicates the request should be tried again after some time.
    /// - Throws: Other ``SignalError``s for other kinds of failures.
    @discardableResult
    public func connect() async throws -> DebugInfo {
        let rawDebugInfo = try await self.tokioAsyncContext.invokeAsyncFunction { promise, tokioAsyncContext in
            withNativeHandle { chatService in
                signal_chat_service_connect_unauth(promise, tokioAsyncContext, chatService)
            }
        }
        return DebugInfo(consuming: rawDebugInfo)
    }

    /// Initiates termination of the underlying connection to the Chat Service. After the service is
    /// disconnected, it will not attempt to automatically reconnect until you call ``connect()``.
    ///
    /// Note: the same instance of `UnauthenticatedChatService` can be reused after `disconnect()` was
    /// called.
    ///
    /// Returns when the disconnection is complete.
    public func disconnect() async throws {
        _ = try await self.tokioAsyncContext.invokeAsyncFunction { promise, tokioAsyncContext in
            withNativeHandle { chatService in
                signal_chat_service_disconnect_unauth(promise, tokioAsyncContext, chatService)
            }
        }
    }

    /// Sends request to the Chat Service over an unauthenticated channel.
    ///
    /// - Throws: ``SignalError/chatServiceInactive(_:)`` if you haven't called ``connect()``
    /// - Throws: Other ``SignalError``s for other kinds of failures.
    /// - SeeAlso: ``sendAndDebug(_:)``
    public func send(_ request: Request) async throws -> Response {
        let internalRequest = try Request.InternalRequest(request)
        let timeoutMillis = request.timeoutMillis
        let rawResponse: SignalFfiChatResponse = try await self.tokioAsyncContext.invokeAsyncFunction { promise, tokioAsyncContext in
            withNativeHandle { chatService in
                internalRequest.withNativeHandle { request in
                    signal_chat_service_unauth_send(promise, tokioAsyncContext, chatService, request, timeoutMillis)
                }
            }
        }
        return try Response(consuming: rawResponse)
    }

    /// Sends request to the Chat Service over an unauthenticated channel.
    ///
    /// In addition to the response, an object containing debug information about the request flow
    /// is returned.
    ///
    /// - Throws: ``SignalError/chatServiceInactive(_:)`` if you haven't called ``connect()``
    /// - Throws: Other ``SignalError``s for other kinds of failures.
    /// - SeeAlso: ``send(_:)``
    public func sendAndDebug(_ request: Request) async throws -> (Response, DebugInfo) {
        let internalRequest = try Request.InternalRequest(request)
        let timeoutMillis = request.timeoutMillis
        let rawResponse: SignalFfiResponseAndDebugInfo = try await self.tokioAsyncContext.invokeAsyncFunction { promise, tokioAsyncContext in
            withNativeHandle { chatService in
                internalRequest.withNativeHandle { request in
                    signal_chat_service_unauth_send_and_debug(promise, tokioAsyncContext, chatService, request, timeoutMillis)
                }
            }
        }
        return (try Response(consuming: rawResponse.response), DebugInfo(consuming: rawResponse.debug_info))
    }
}
